### PR TITLE
Avoid calling `Topic::new` for every received Message

### PR DESCRIPTION
**Pull request by Swatinem**
That function is interning the topic name, and thus aquiring a Mutex. If we keep a list of subscribed topics locally, we avoid that and can also guard against errors.

### CODE REVIEW

File: rust_snuba/rust_arroyo/src/backends/kafka/mod.rs [hunk 1]
```diff
@@ -78,8 +78,14 @@ impl KafkaConsumerState {
...
-fn create_kafka_message(msg: BorrowedMessage) -> BrokerMessage<KafkaPayload> {
-    let topic = Topic::new(msg.topic());
+fn create_kafka_message(topics: &[Topic], msg: BorrowedMessage) -> BrokerMessage<KafkaPayload> {
+    let topic = msg.topic();
+    let topic = *topics
+        .iter()
+        .find(|t| t.as_str() == topic)
+        .unwrap_or_else(|| {
+            panic!("Received message for topic `{topic}` that we never subscribed to");
+        });
```
Comment (loewenheim) [inline rust_snuba/rust_arroyo/src/backends/kafka/mod.rs:88]: `unwrap_or_else` with a `panic` inside is more naturally written with `let else`, I think.

File: rust_snuba/rust_arroyo/src/backends/kafka/mod.rs [hunk 2]
```diff
@@ -328,9 +335,11 @@ impl<C: AssignmentCallbacks> KafkaConsumer<C> {
 
         let topic_str: Vec<&str> = topics.iter().map(|t| t.as_str()).collect();
         consumer.subscribe(&topic_str)?;
+        let topics = topics.to_owned();
```
Comment (loewenheim) [inline rust_snuba/rust_arroyo/src/backends/kafka/mod.rs:338]: Normally I'd advocate for passing the already-owned value in, but it'd probably cause a bunch of changes for not a lot of benefit.

**Review by untitaker — APPROVED** [review]
please add some code comments for why we really do this kind of validation (in order to avoid calling Topic::new), otherwise lgtm

File: rust_snuba/rust_arroyo/src/backends/kafka/mod.rs [hunk 3]
```diff
@@ -78,10 +78,15 @@ impl KafkaConsumerState {
     }
 }
 
-fn create_kafka_message(msg: BorrowedMessage) -> BrokerMessage<KafkaPayload> {
-    let topic = Topic::new(msg.topic());
+fn create_kafka_message(topics: &[Topic], msg: BorrowedMessage) -> BrokerMessage<KafkaPayload> {
+    let topic = msg.topic();
+    // NOTE: We avoid calling `Topic::new` here, as that uses a lock to intern the `topic` name.
+    // As we only ever expect one of our pre-defined topics, we can also guard against Broker errors.
+    let Some(topic) = topics.iter().find(|t| t.as_str() == topic) else {
```
Comment (loewenheim) [inline rust_snuba/rust_arroyo/src/backends/kafka/mod.rs:85]:
```suggestion
    let Some(&topic) = topics.iter().find(|t| t.as_str() == topic) else {
```

File: rust_snuba/rust_arroyo/src/backends/kafka/mod.rs [hunk 4]
```diff
@@ -78,10 +78,15 @@ impl KafkaConsumerState {
...
+        panic!("Received message for topic `{topic}` that we never subscribed to");
+    };
     let partition = Partition {
-        topic,
+        topic: *topic,
```
Comment (loewenheim) [inline rust_snuba/rust_arroyo/src/backends/kafka/mod.rs:89]:
```suggestion
        topic,
```

**Pull request merged**